ctk/common/src/main/kotlin/org/codice/compliance/utils/schema/SchemaValidator.kt

```diff
@@ -0,0 +1,111 @@
+import java.io.ByteArrayOutputStream
+
+class SchemaValidator {
+    companion object {
+
+        private const val JAXP_SCHEMA_LANGUAGE =
+                "http://java.sun.com/xml/jaxp/properties/schemaLanguage"
+        private const val JAXP_SCHEMA_SOURCE =
+                "http://java.sun.com/xml/jaxp/properties/schemaSource"
+        private const val EXTERNAL_SCHEMA_LOCATION =
+                "http://apache.org/xml/properties/schema/external-schemaLocation"
+
+        private val schemas = arrayOf(
+                "saml-schema-assertion-2.0.xsd",
+                "saml-schema-protocol-2.0.xsd"
+        )
+
+        private val schemaLocationOverrides = mapOf(
+                "http://www.w3.org/2000/09/xmldsig#" to "xmldsig-core-schema.xsd",
+                "http://www.w3.org/2001/04/xmlenc#" to "xenc-schema.xsd"
+        )
+
+        fun validateSAMLMessage(saml: Node) {
+            // Convert SAML to InputStream
+            val outputStream = ByteArrayOutputStream()
+            val xmlSource = DOMSource(saml)
+            val outputTarget = StreamResult(outputStream)
+            TransformerFactory.newInstance().newTransformer().transform(xmlSource, outputTarget)
+            val samlInputStream = ByteArrayInputStream(outputStream.toByteArray())
+
+            // Create array of schema InputStreams
+            val schemaInputStreams = schemas
+                    .map { this::class.java.classLoader.getResourceAsStream(it) }
+                    .toTypedArray()
+
+            // Create DocumentBuilder for schema
+            val dbf = DocumentBuilderFactory.newInstance()
+            dbf.isNamespaceAware = true
+            dbf.isValidating = true
+            dbf.setAttribute(JAXP_SCHEMA_LANGUAGE, XMLConstants.W3C_XML_SCHEMA_NS_URI)
+            dbf.setAttribute(JAXP_SCHEMA_SOURCE, schemaInputStreams)
+            dbf.setAttribute(EXTERNAL_SCHEMA_LOCATION, schemaLocationOverrides
+                    .map { it.key to this::class.java.classLoader.getResource(it.value).toExternalForm() }
+                    .map { "${it.first} ${it.second}" }
+                    .joinToString("\n")
+            )
+            dbf.setAttribute(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "file")
+
+            // Validate and report errors
+            val db = dbf.newDocumentBuilder()
+            db.setErrorHandler(SAMLErrorHandler())
+            db.parse(samlInputStream)
+        }
+    }
+
+    private class SAMLErrorHandler : ErrorHandler {
+
+        private fun getParseExceptionInfo(spe: SAXParseException): String {
+            var systemId: String? = spe.systemId
+            if (systemId == null) {
+                systemId = "null"
+            }
+
+            return "URI=" + systemId + " Line=" + spe.lineNumber + ": " + spe.message
+        }
+
+        @Throws(SAXException::class)
+        override fun warning(spe: SAXParseException) {
+            Log.debugWithSupplier { "Warning: " + getParseExceptionInfo(spe) }
```

Comment:
Do you know what an example of a warning would be in this case? We want to be pretty strict about them following the schema.